File: 단지번호붙이기.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>

using namespace std;

const int MAX=25;
int g[MAX+1][MAX+1];
bool c[MAX+1][MAX+1];
vector<int> numOfHouse;
int n;
int dx[] = {-1,1,0,0};
int dy[] = {0,0,1,-1};

int dfs(int x,int y){
  c[x][y]=true;
  int cnt =1;
  for(int i=0; i<4; i++){
    int nx = x+ dx[i];
    int ny = y +dy[i];
    if(nx<0||nx>=n||ny<0||ny>=n) continue;
    if(g[nx][ny] == 0 ) continue;
    if(c[nx][ny]) continue;
    cnt+=dfs(nx,ny);
  }
  return cnt;
}

int main(){
  cin>>n;
  for(int i=0; i<n; i++){
    string temp;
    cin>>temp;
    for(int j=0; j<n; j++){
      g[i][j] = temp[j]-'0';
    }
  }

  int num =0;
  for(int i=0; i<n; i++){
    for(int j=0; j<n; j++){
      if(g[i][j] && !c[i][j]){
        int num = dfs(i,j);
        numOfHouse.push_back(num);
      }
    }
  }

  sort(numOfHouse.begin(),numOfHouse.end());
  cout<<numOfHouse.size()<<'\n';
  for(int i=0; i<numOfHouse.size(); i++){
    cout<<numOfHouse[i]<<'\n';
  }
}

File: 단지번호붙이기.py
from collections import deque

n=int(input())
g=[list(map(int,list(input()))) for _ in range(n)]
c=[[False]*n for _ in range(n)]
numOfHouse = []
dx=(-1,1,0,0)
dy=(0,0,-1,1)

def bfs(sx,sy):
  c[sx][sy]= True
  q= deque()
  q.append((sx,sy))
  cnt=1
  while q:
    x,y= q.popleft()
    for i in range(4):
      nx=x+dx[i]
      ny=y+dy[i]
      if nx<0 or nx>=n or ny<0 or ny>=n: continue
      if g[nx][ny]==0 : continue
      if c[nx][ny]: continue
      c[nx][ny]=True
      cnt+=1
      q.append((nx,ny))
  return cnt

for i in range(n):
  for j in range(n):
    if g[i][j] == 1 and c[i][j]==False :
      ret=bfs(i,j)
      numOfHouse.append(ret)
numOfHouse.sort()

print(len(numOfHouse))
for i in range(len(numOfHouse)):
  print(numOfHouse[i])
      
      